### PR TITLE
cop: avoid handling requests multiple times

### LINKED ISSUE #6676: TPC-H q8 & q21 results are wrong on master
### What version of TiKV are you using?

58ed7e14a6f1a39c0ae50260547d78648b5f60c8

### What operating system and CPU are you using?

```
[tidb@localhost tmp]$ lscpu
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                40
On-line CPU(s) list:   0-39
Thread(s) per core:    2
Core(s) per socket:    10
Socket(s):             2
NUMA node(s):          2
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 79
Model name:            Intel(R) Xeon(R) CPU E5-2630 v4 @ 2.20GHz
Stepping:              1
CPU MHz:               2873.937
CPU max MHz:           3100.0000
CPU min MHz:           1200.0000
BogoMIPS:              4400.04
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              256K
L3 cache:              25600K
NUMA node0 CPU(s):     0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30,32,34,36,38
NUMA node1 CPU(s):     1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31,33,35,37,39
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch epb cat_l3 cdp_l3 intel_pt ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm rdt_a rdseed adx smap xsaveopt cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts spec_ctrl intel_stibp
```

### Steps to reproduce

Use https://github.com/pingcap/tidb-bench/tree/master/tpch to execute q8 and q21, scale factor was 50 here.

### What did you expect?

```
$ mysql -h 127.0.0.1 -uroot -P4000 -D TPCH50  < 08.sql

# expect
o_year	mkt_share
1995	0.03944761
1996	0.03866107
```

### What did happened?

I executed 4 times, results were below:

```
# got(1st)

o_year	mkt_share
1995	0.00000000
1996	0.00000000

# got(2nd)

# got(3td)
o_year	mkt_share
1995	0.03081039
1996	0.14589505

# got(4th)
ERROR 1105 (HY000) at line 4: indexHashJoinInnerWorker.doJoinUnordered failed
```

You can see different results when execute q8.

### PR DESCRIPTION
**Pull request by sticnarf**
###  What have you changed?

In #6582, a request in execution which fails to get a semaphore permit can be cancelled and rerun until a semaphore permit is acquired. However, the problem is that the coprocessor executors have inner state, so it is incorrect to rerun the request handler.

This PR changes the mechanism. The semaphore is passed into the runner. If the execution limit exceeds, the runner blocks itself through the semaphore. Now, the handler is not used more than once.

###  What is the type of the changes?

- Bugfix (a change which fixes an issue)

###  How is the PR tested?

- Unit test
- Manual test

I reproduced the bug according to #6676 in master. In this fixed version, the problem does not exist.

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

Need to add this PR to the unified thread pool changelog as well.

###  Does this PR affect `tidb-ansible`?

No

###  Refer to a related PR or issue link (optional)

Should fix #6676

### CODE REVIEW

File: src/coprocessor/endpoint.rs [hunk 1]
```diff
@@ -203,16 +204,19 @@ impl<E: Engine> Endpoint<E> {
...
                     );
-                    dag::build_handler(
+                    dag::DagHandlerBuilder::new(
                         dag,
                         ranges,
                         store,
-                        data_version,
                         req_ctx.deadline,
                         batch_row_limit,
                         is_streaming,
```
Comment (iosmanthus) [inline src/coprocessor/endpoint.rs]: Seems `is_streaming` is another option for `DagHandlerBuilder`, could we specify a builder for it?

Comment (sticnarf) [reply src/coprocessor/endpoint.rs]: I expose all compulsary options only to the `new` function. Currently it is not uncessary to provide a modifier for these options.

**Comment by mahjonp** [comment]
/bench +tpch
```
workload:
  queries: 08.sql 21.sql
  repeat_time: 3
```

**Pull request merged**